sycl/plugins/opencl/pi_opencl.cpp

```diff
@@ -516,29 +542,62 @@ pi_result piextGetDeviceFunctionPointer(pi_device device, pi_program program,
     return cast<pi_result>(ret_err);
   }
 
-  using FuncT =
-      cl_int(CL_API_CALL *)(cl_device_id, cl_program, const char *, cl_ulong *);
-
-  // TODO: add check that device supports corresponding extension
-  FuncT func_ptr =
-      reinterpret_cast<FuncT>(clGetExtensionFunctionAddressForPlatform(
-          cast<cl_platform_id>(platform), "clGetDeviceFunctionPointerINTEL"));
-  // TODO: once we have check that device supports corresponding extension,
-  // we can insert an assertion that func_ptr is not nullptr. For now, let's
-  // just return an error if failed to query such function
-  // assert(
-  //     func_ptr != nullptr &&
-  //     "Failed to get address of clGetDeviceFunctionPointerINTEL function");
-
-  if (!func_ptr) {
-    if (function_pointer_ret)
-      *function_pointer_ret = 0;
-    return PI_INVALID_DEVICE;
+  // Get device supported extensions
+  size_t extension_size;
+  clGetDeviceInfo(cast<cl_device_id>(device), CL_DEVICE_EXTENSIONS, 0, nullptr,
+                  &extension_size);
+  char *extensions = new char[extension_size];
+  clGetDeviceInfo(cast<cl_device_id>(device), CL_DEVICE_EXTENSIONS,
+                  extension_size, extensions, nullptr);
+
+  // Check if clGetDeviceFunctionPointer is in list of extensions
+  if (is_in_separated_string(extensions, ' ',
+                             "clGetDeviceFunctionPointerINTEL")) {
+    delete[] extensions;
+
+    using FuncT = cl_int(CL_API_CALL *)(cl_device_id, cl_program, const char *,
+                                        cl_ulong *);
+
+    FuncT func_ptr =
+        reinterpret_cast<FuncT>(clGetExtensionFunctionAddressForPlatform(
+            cast<cl_platform_id>(platform), "clGetDeviceFunctionPointerINTEL"));
```

Comment:
Please use getExtFuncFromContext that implements caching of the extensions

Reply:
I have changed this to use `getExtFuncFromContext` so it the function should be cached